core/arch/arm/plat-hikey/spi_test.c

```diff
@@ -48,45 +72,57 @@
  */
 void spi_test(void)
 {
-	struct pl061_data platform_pl061_data;
+	struct pl061_data platform_pl061_data __maybe_unused;
 	struct pl022_data platform_pl022_data;
-	vaddr_t gpio6_base = nsec_periph_base(GPIO6_BASE);
+	vaddr_t __maybe_unused gpio6_base = nsec_periph_base(GPIO6_BASE);
 	vaddr_t spi_base = nsec_periph_base(SPI_BASE);
 	uint8_t tx[3] = {0x01, 0x80, 0x00};
 	uint8_t rx[3] = {0};
 	size_t i, j, len = 3;
+	bool local_cs = false;
 
-	DMSG("gpio6_base: 0x%" PRIxVA "\n", gpio6_base);
-	DMSG("spi_base: 0x%" PRIxVA "\n", spi_base);
-
-	DMSG("configure GPIO\n");
-	pl061_init(&platform_pl061_data);
-	pl061_register(gpio6_base, 6);
+	if (!local_cs) {
+		DMSG("gpio6_base: 0x%" PRIxVA "\n", gpio6_base);
+		DMSG("configure GPIO\n");
+		pl061_init(&platform_pl061_data);
+		pl061_register(gpio6_base, 6);
+		DMSG("enable software mode control for chip select\n");
+		pl061_set_mode_control(GPIO6_2, PL061_MC_SW);
 
-	DMSG("enable software mode control for chip select\n");
-	pl061_set_mode_control(GPIO6_2, PL061_MC_SW);
-
-	DMSG("mask/disable interrupt for chip select\n");
-	platform_pl061_data.chip.ops->set_interrupt(GPIO6_2,
-						GPIO_INTERRUPT_DISABLE);
+		platform_pl022_data.cs_data.gpio_data.chip =
+			&platform_pl061_data.chip;
+		platform_pl022_data.cs_data.gpio_data.pin_num = GPIO6_2;
+		platform_pl022_data.b_cs_callback = false;
+	} else {
+		DMSG("set CS callback\n");
+		/*
+		 * NULL can be registered, in which case you can pulse CS as
+		 * desired below!
+		 */
+		platform_pl022_data.cs_data.cs_cb = spi_cs_callback;
+		platform_pl022_data.b_cs_callback = true;
+	}
 
+	DMSG("spi_base: 0x%" PRIxVA "\n", spi_base);
 	DMSG("configure SPI\n");
-	platform_pl022_data.gpio = &platform_pl061_data.chip;
 	platform_pl022_data.base = spi_base;
-	platform_pl022_data.cs_gpio_base = gpio6_base;
 	platform_pl022_data.clk_hz = SPI_CLK_HZ;
 	platform_pl022_data.speed_hz = SPI_10_KHZ;
-	platform_pl022_data.cs_gpio_pin = GPIO6_2;
 	platform_pl022_data.mode = 0;
 	platform_pl022_data.data_size_bits = 8;
-	platform_pl022_data.loopback = true;
+	platform_pl022_data.b_loopback = true;
 
 	pl022_init(&platform_pl022_data);
 	platform_pl022_data.chip.ops->configure(&platform_pl022_data.chip);
 	platform_pl022_data.chip.ops->start(&platform_pl022_data.chip);
 
+	/* If cs_cb = NULL, CS can be enabled here */
+
 	for (j = 0; j < 20; j++) {
 		DMSG("SPI test loop: %zu\n", j);
+
+		/* If cs_cb = NULL, CS can be enabled here */
+
```

Comment:
These comments follow the one above that says "NULL can be registered, in which case you can pulse CS as desired below". They're for the use case where start and end handlers don't control the CS, i.e. when `b_cs_callback` is true.

Perhaps I was trying to show too much in one function. The other option is to have 3 spi test functions where:
i. `b_cs_callback` is false, i.e. start and end handlers don't control the CS
ii. `b_cs_callback' is true and `cs_cb` is not NULL
iii. `b_cs_callback' is true and `cs_cb` is NULL
Maybe this way it can illustrate better how the code differs between use cases?